Fix widget sizing on current Gtk+ 3.22 branch

heftig opened the pull request:
[Gtk recently started emitting configure events when repositioning child windows](https://git.gnome.org/browse/gtk+/commit/?h=gtk-3-22&id=12579fe71b3b8f79eb9c1b80e429443bcc437dd0). Events for the form's child windows bubble up the widget hierarchy, get caught by our configure event handler, and treated as if they were reporting the size of the form window. [As a result, the shell was sized using far too small dimensions](https://bugs.archlinux.org/task/51509). ([bgo](https://bugzilla.gnome.org/show_bug.cgi?id=773387)) This change will be part of Gtk+ 3.22.2.

Solve this by checking whether the configure event we get is really for the form widget.

heftig committed:
Fix widget sizing on current Gtk+ 3.22 branch

Gtk recently started emitting configure events when repositioning child
windows[1]. Events for the form's child windows bubble up the widget
hierarchy, get caught by our configure event handler, and treated as if
they were reporting the size of the form window. As a result, the shell
was sized using far too small dimensions[2][3]. This change will be part
of Gtk+ 3.22.2.

Solve this by checking whether the configure event we get is really for
the form widget.

[1]: https://git.gnome.org/browse/gtk+/commit/?h=gtk-3-22&id=12579fe71b3b8f79eb9c1b80e429443bcc437dd0
[2]: https://bugs.archlinux.org/task/51509
[3]: https://bugzilla.gnome.org/show_bug.cgi?id=773387

nuko8 commented:
> As a result, the shell was sized using far too small dimensions2. 

Thinking about the current implementation of our GtkForm, I wonder if that really happend to gvim with the said change in GdkWindow.  Could you tell us more about the bug itself?

> This change will be part of Gtk+ 3.22.2.

Is that a final decision?

heftig commented:
> Thinking about the current implementation of our GtkForm, I wonder if that really happend to gvim with the said change in GdkWindow. Could you tell us more about the bug itself?

The symptom is a nearly empty window—most obviously, the main text widget is missing. While resizing the window, you could occasionally see it get drawn for a frame, but it immediately disappears again.

The initial reporter already determined the commit which was causing the problem. I poked around a bit and saw Gtk complaining about zero sizes being set. `G_DEBUG=fatal-warnings` pointed me to the configure-event handler, and eventually I added a printf there to print out the width and height.

I noticed that on each resize it fired multiple times, the first time having the correct size and the subsequent times being too small, from a slightly smaller size to much smaller sizes, most of them 1x1.

I suspected that the handler was seeing configure events for child widgets, and added the check to disregard them. That solved the issue. I also rechecked with Gtk 3.22.1 to verify it still works.

> Is that a final decision?

Probably. I'm still waiting for a verdict on https://bugzilla.gnome.org/show_bug.cgi?id=773387 .

The change was made to fix https://bugzilla.gnome.org/show_bug.cgi?id=771320 and https://bugzilla.gnome.org/show_bug.cgi?id=767713 . From my limited understanding of GTK it seems to be sound.

heftig commented:
Seems this isn't a complete fix yet, as when using gnome-shell's tiling or maximizing the window, the widget isn't resized.

heftig commented:
Entering visual mode even once seems to unbreak the resizing on maximize.

heftig committed:
Further fix widget sizing

Follow-up by ofourdan in order to fix resizing on maximization. This
makes the drawing area ignore configure events that weren't explicitly
sent (such as from GtkForm).

heftig commented:
ftr

> <ofourdan> note, I am not saying this is the right fix, but for now I don't know how to address the issue, in one case we want the configure events from gdk_window_move_resize_internal() because otherwise widgets such as GtkClutterEmbed which use a subsurface in Wayland do not /know/ of the new size/position, in the other case it seems gvim picks the wrong events as the gdkwindow comes after the size_allocate
> <ofourdan> and still have the "old" size
> <ofourdan> I would be willing to revert commit 12579fe but then I dunno how to fix the two issue this commit was to address...
> 
> <ofourdan> what happens is: configure_native_child() queue an event with the old size, gtkdrawingarea emits a configure event with the new size, then gtk dequeue the events and re-emit a configure with the old size
> <ofourdan> which makes me think that the fix in gtk is broken and needs to be reverted
> <ofourdan> yet I am still searching for another solution to the problems it was supposed to address

nuko8 commented:
@heftig 

Contrary to the impression I got from the words (of a gtk dev, I presume):

> (...) I am not saying this is the right fix
> (...) I would be willing to revert commit 12579fe
> (...) which makes me think that the fix in gtk is broken and needs to be reverted yet I am still searching for another solution to the problems it was supposed to address,

the change in question has actually been included in 3.22.2.

I installed 3.22.2 last Sunday and confirmed that it broke gvim as you reported, and that your patch worked for me.

That said, I'm still not convinced whether or not the code newly added to around line 3102 of gui_gtk_x11.c is a right fix.  (Note that I leave this comment for record, not that I'm trying to ask you to address this issue right now.)

We could wait for him, ofourdan, coming up with the right solution; however, once the library is updated, it is hard for the users to make it revert safely.  Therefore, it would be better for us to do something about that if there would be something we could do rather than waiting for the right fix done in gtk.

Accordingly, I would think it's appropriate to merge your patch into the master if you could enclose the two blocks of the code with preprocessor directives like this:

``` c
#if GTK_CHECK_VERSION(3,22,2)
    <your new code>
#endif 
```

Please note that, if, necessary, add the appropriate number of spaces between `#` and a directive to show the nest level.  By that guard, we wouldn't need to worry about any possible regression against the library older than 3.22.2.

Thank you for the report and the patch.

@brammool 
After the version check directives are added by @heftig, please consider the inclusion of his patch.  I've been using a gvim applied with the patch since last Sunday and found no issue so far with it.  Actually, using that patched gvim, I prepared another patch to address other issues on 3.22.2.  I'll send you my patch after @heftig 's revised patch is included, then making my patch rebased on that.

heftig commented:
I'm actually reasonably confident these changes won't break on older GTK3 versions.

nuko8 commented:
In this particular case, simply adding four lines of code to the patch 100% guarantees no break, and by doing that, no break can be easily achieved and verified even by people who are not necessarily GTK experts, which is much more important to maintain the codebase after your patch might be included than accepting changes on the basis of someone's confidence which no one can verify.

heftig commented:
Very well; I chose to go with a runtime check, though, since this isn't an ABI change.

nuko8 commented:
The reason is not strong enough to justify introducing some overhead which is caused by the new runtime check to all the gvim linked against <= 3.22.1, which overhead is obviously unnecessary and avoidable.

Also , it looks strange and awkward that `gtk_check_version(3,22,2)` is enclosed by `#if GTK_CHECK_VERSION(3,0,0) .. #endif`.  That (with a wrong indentation) would confuse people when they visit the code for the purpose of maintenance or enhancement months or years later after a merge takes place.

Furthermore, I noticed that the patch didn't work for the case where one tries to do `:vsplit` in fullscreen with scrollbars enabled.

heftig commented:
No, the split makes sense.

If you build gvim against GTK 2 you **can't** run it with GTK 3.22.2, so the new code should not be present in this case.

If you build gvim against GTK 3.0.0 you **can** run it with GTK 3.22.2, so the new code should be present. But as you asked we should not risk altering the behavior of GTK < 3.22.2, so the runtime check is needed.

Anyway, this has been the extent of time and energy I'm willing to spend on this bug.

nuko8 commented:
> If you build gvim against GTK 3.0.0 you can run it with GTK 3.22.2, so the new code should be present. But as you asked we should not risk altering the behavior of GTK < 3.22.2, so the runtime check is needed.

That depends on the characteristics of the dynamic loader in use, and hence on the OS, whether or not a binary which is built like that works well with a version of the library higher than that against which it was linked at build time.  On some systems, the dynamic loader rejects linking and tells the user about version mismatch.

In addition, it is the fact, at least from my experience during the gtk3 gui development,  that, once a function is deprecated in gtk3, it is not maintained well and often stop working at runtime even if it was built without problem.  That's why the gtk3 code involves many of `#if GTK_CHEK_VERSION()`.  Those are non-trivial efforts to make the gui work well for every still-used version of gtk3.   Don't break that.

Accordingly, it is strongly discouraged for people to run a binary built with an old gtk3 with a newer gtk3.  No one can guarantee it works. 

Therefore, any runtime check doesn't make sense at all.  Gtk 3 is different from other ordinary library with this respect because of its drastic changes happening even between micro version number updates.

ismail commented:
I applied the two patches on top of vim 8.0.45 (with gtk+ 3.22.2) and gvim window still does not render properly, getting the following warnings on console:

```
(gvim:26226): Gtk-CRITICAL **: gtk_widget_set_size_request: assertion 'width >= -1' failed
*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug


(gvim:26226): Gtk-CRITICAL **: gtk_widget_set_size_request: assertion 'width >= -1' failed
*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug


(gvim:26226): Gtk-CRITICAL **: gtk_widget_set_size_request: assertion 'width >= -1' failed
*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug


(gvim:26226): Gtk-CRITICAL **: gtk_widget_set_size_request: assertion 'width >= -1' failed
```

nsuke commented on src/gui_gtk_x11.c:
```diff
@@ -3099,6 +3099,11 @@ drawarea_configure_event_cb(GtkWidget	      *widget,
 	return TRUE;
 
+#if GTK_CHECK_VERSION(3,0,0)
+    if (gtk_check_version(3, 22, 2) && event->send_event == FALSE)
```
The function version of `gtk_check_version` returns NULL when compatible,
unlike the macro counterpart that returns TRUE.
Inverting the check (prepend a `!`) made this patch functional again for me.

nsuke commented on src/gui_gtk_x11.c:
```diff
@@ -3099,6 +3099,11 @@ drawarea_configure_event_cb(GtkWidget	      *widget,
     if (event->width == cur_width && event->height == cur_height)
 	return TRUE;
 
+#if GTK_CHECK_VERSION(3,0,0)
```
FYI, we're in an exactly the same `#if` clause here.

heftig commented:
Gtk 3.22.4 should have fixed this.
